### ISSUE TITLE
Can I save some values in training process by optuna?

### ISSUE THREAD

**Issue by DRJYYDS** [optuna/optuna#3455]
### Environment

None

### Description

Sometimes we need save some training value like val_acc/train_acc, but optuna usually just save the training metric and params, so is there any method to save those training value by using optuna?

### Error messages, stack traces, or logs

```shell
None
```

### Steps to reproduce

None

**Comment by nzw0301** [comment]
Yes, https://github.com/optuna/optuna/issues/3446#issuecomment-1090094284

**Comment by DRJYYDS** [comment]
> Refer

But it this seems only save one value for on variable, the situation is that I want to save a array of training_loss/val_loss, ant visualize it after training, I know trial.report can do this, but it only can save one variable.

Btw the trial.report is pretty convenient, and it even can be used in optuna-dashboard to show it automatically, is there some api like trial.report allow us to save and visualize intermediate variable in training process?

Thanks for your reply

**Comment by DRJYYDS** [comment]
> 

I got a solution now... Actually optuna has integrated tensorboard, I can write to tensorboard(with the rundir) by my own, and set the callback as tensorboard, But I think it's easy to provide an api in optuna and convenient for users, maybe you guys can add such an api for users.

**Comment by nzw0301** [comment]
I'm sorry, I couldn't understand what is the proposal in the previous comment. Could you explain a little bit more? Ideally, minimal reproducible codes are much appreciated.

**Comment by DRJYYDS** [comment]
> I'm sorry, I couldn't understand what is the proposal in the previous comment. Could you explain a little bit more? Ideally, minimal reproducible codes are much appreciated.
`import random
import optuna
from optuna.integration.tensorboard import TensorBoardCallback
import tensorflow as tf
import os
dirname = "logs_test"

def objective(trial: optuna.trial.Trial) -> float:
    x = trial.suggest_int("x", 16, 32)
    y = trial.suggest_float("y", 0.1, 0.2)
    z = trial.suggest_float("z", 0.1, 0.2)
    run_name = "trial-%d" % trial.number
    run_dir = os.path.join(dirname, run_name) # This is the dir used in tensorboard_callback of optuna
    with tf.summary.create_file_writer(run_dir).as_default():
      for i in range(10):
        tf.summary.scalar("trainloss", random.random(), step=i)
        tf.summary.scalar("testloss", random.random(), step=i)
        tf.summary.scalar("valloss", random.random(), step=i)
    target = y = -(x-2)**2 + y + z**3 + 2*z - 6*z**2
    return target


tensorboard_callback = TensorBoardCallback(dirname = dirname, metric_name="target")
study = optuna.create_study(direction="maximize")
study.optimize(objective, n_trials=10, timeout=600, callbacks=[tensorboard_callback])`

## Result:
![image](https://user-images.githubusercontent.com/102657334/162616214-bb98773b-0b05-4278-b5b5-ef514b36e00e.png)
This allow us to save and visualize the intermediate value and hyper-params in the traning process by tensorboard.
But it's a litlle bit trouble, cause you need calculate the run_dir in the training function.
If optuna provide an api related to this(just wrap the tensorboard), we can easily to save and visulize the intermediate value, that would be great. If optuna-dashboard support this feature, that would be awesome!

**Comment by DRJYYDS** [comment]
> I'm sorry, I couldn't understand what is the proposal in the previous comment. Could you explain a little bit more? Ideally, minimal reproducible codes are much appreciated.

## Code
`import random
import optuna
from optuna.integration.tensorboard import TensorBoardCallback
import tensorflow as tf
import os
dirname = "logs_test"

def objective(trial: optuna.trial.Trial) -> float:
    x = trial.suggest_int("x", 16, 32)
    y = trial.suggest_float("y", 0.1, 0.2)
    z = trial.suggest_float("z", 0.1, 0.2)
    run_name = "trial-%d" % trial.number
    run_dir = os.path.join(dirname, run_name) # This is the dir used in tensorboard_callback of optuna
    with tf.summary.create_file_writer(run_dir).as_default():
      for i in range(10):
        tf.summary.scalar("trainloss", random.random(), step=i)
        tf.summary.scalar("testloss", random.random(), step=i)
        tf.summary.scalar("valloss", random.random(), step=i)
    target = y = -(x-2)**2 + y + z**3 + 2*z - 6*z**2
    return target


tensorboard_callback = TensorBoardCallback(dirname = dirname, metric_name="target")
study = optuna.create_study(direction="maximize")
study.optimize(objective, n_trials=10, timeout=600, callbacks=[tensorboard_callback])`

## Result
![image](https://user-images.githubusercontent.com/102657334/162616405-d8be626d-29b1-4c41-8c50-0c15e1d81ab7.png)

This allow us to save and visualize the intermediate value and hyper-params in the traning process by tensorboard.
But it's a litlle bit trouble, cause you need calculate the run_dir in the training function.
If optuna provide an api related to this(just wrap the tensorboard), we can easily to save and visulize the intermediate value, that would be great. If optuna-dashboard support this feature, that would be awesome!

**Comment by nzw0301** [comment]
Thank you for your clarification with examples!

### TFBoard

[`MLflowCallback`](https://optuna.readthedocs.io/en/latest/reference/generated/optuna.integration.MLflowCallback.html) (and wandb as well soon) has similar functionality to what you want. So I suppose we could provide similar functionality with TFboard callback.

### Optuna-dashboard
The development of Optuna-dashboard is in a different repo: https://github.com/optuna/optuna-dashboard. Could you create an issue there?

**Comment by DRJYYDS** [comment]
> Thank you for your clarification with examples!
> 
> ### TFBoard
> [`MLflowCallback`](https://optuna.readthedocs.io/en/latest/reference/generated/optuna.integration.MLflowCallback.html) (and wandb as well soon) has similar functionality to what you want. So I suppose we could provide similar functionality with TFboard callback.
> 
> ### Optuna-dashboard
> The development of Optuna-dashboard is in a different repo: https://github.com/optuna/optuna-dashboard. Could you create an issue there?

The mlflow seems can't show the curve(just like the train_loss curve in my example), it just log a value and no visulization, is my thought right? I think in visulization, tensorboard is more powerful...

Thanks for your reply, I will create an issue in optuna-dashboard later.

**Comment by nzw0301** [comment]
How about this?

```python
import random
import optuna
from optuna.integration.mlflow import MLflowCallback
import mlflow
import os


mlflc = MLflowCallback(
    tracking_uri="http://127.0.0.1:5000/",
    metric_name="my metric score",
)


@mlflc.track_in_mlflow()
def objective(trial: optuna.trial.Trial) -> float:
    x = trial.suggest_int("x", 16, 32)
    y = trial.suggest_float("y", 0.1, 0.2)
    z = trial.suggest_float("z", 0.1, 0.2)

    for i in range(10):
        mlflow.log_metric("trainloss", random.random(), step=i)
        mlflow.log_metric("testloss", random.random(), step=i)
        mlflow.log_metric("valloss", random.random(), step=i)

    target = y = -(x-2)**2 + y + z**3 + 2*z - 6*z**2
    return target


study = optuna.create_study(direction="maximize")
study.optimize(objective, n_trials=10)
```

and 
```sh
mlflow uI # go to http://127.0.0.1:5000/
```

Then we can see visualization on http://127.0.0.1:5000/:


<img width="1821" alt="Screenshot 2022-04-10 at 23 10 45" src="https://user-images.githubusercontent.com/7121753/162623485-14f46001-c38a-4c5c-bf01-4c2d0136c06c.png">


<img width="1843" alt="Screenshot 2022-04-10 at 23 11 11" src="https://user-images.githubusercontent.com/7121753/162623482-172b9dde-21a3-4e61-8c1d-3aa7432d6bf1.png">

**Comment by DRJYYDS** [comment]
> How about this?
> 
> ```python
> import random
> import optuna
> from optuna.integration.mlflow import MLflowCallback
> import mlflow
> import os
> 
> 
> mlflc = MLflowCallback(
>     tracking_uri="http://127.0.0.1:5000/",
>     metric_name="my metric score",
> )
> 
> 
> @mlflc.track_in_mlflow()
> def objective(trial: optuna.trial.Trial) -> float:
>     x = trial.suggest_int("x", 16, 32)
>     y = trial.suggest_float("y", 0.1, 0.2)
>     z = trial.suggest_float("z", 0.1, 0.2)
> 
>     for i in range(10):
>         mlflow.log_metric("trainloss", random.random(), step=i)
>         mlflow.log_metric("testloss", random.random(), step=i)
>         mlflow.log_metric("valloss", random.random(), step=i)
> 
>     target = y = -(x-2)**2 + y + z**3 + 2*z - 6*z**2
>     return target
> 
> 
> study = optuna.create_study(direction="maximize")
> study.optimize(objective, n_trials=10)
> ```
> 
> and
> 
> ```shell
> mlflow uI # go to http://127.0.0.1:5000/
> ```
> 
> Then we can see visualization on http://127.0.0.1:5000/:
> 
> <img alt="Screenshot 2022-04-10 at 23 10 45" width="1821" src="https://user-images.githubusercontent.com/7121753/162623485-14f46001-c38a-4c5c-bf01-4c2d0136c06c.png">
> 
> <img alt="Screenshot 2022-04-10 at 23 11 11" width="1843" src="https://user-images.githubusercontent.com/7121753/162623482-172b9dde-21a3-4e61-8c1d-3aa7432d6bf1.png">

Got it. Looks pretty good!
Thanks for your reply! I will close the issue.

**Comment by kailkuhn** [comment]
is there a possibility to pass multiple arguments to the objective function when using the @mlflc.track_in_mlflow() decorator? It hasn't really worked for me.
